File: POO/Constructores_Clases/Abstract class/js/Empleado.js
class Empleado {
    constructor() {
        if (this.constructor === Empleado) {
            throw new Error("No se puede instanciar una clase abstracta");
        }
    }

    trabajar() {
        throw new Error("Método abstracto debe ser implementado");
    }

    calcularSalario() {
        throw new Error("Método abstracto debe ser implementado");
    }

    mostrar() {
        throw new Error("Método abstracto debe ser implementado");
    }
}

class Desarrollador extends Empleado {
    constructor(nombre, lenguaje, salario) {
        super();
        this.nombre = nombre;
        this.lenguaje = lenguaje;
        this.salario = salario;
    }

    // Constructor de copia
    static from(other) {
        return new Desarrollador(other.nombre, other.lenguaje, other.salario);
    }

    trabajar() {
        return `Programando en ${this.lenguaje}`;
    }

    calcularSalario() {
        return this.salario;
    }

    mostrar() {
        console.log(`Desarrollador: ${this.nombre}`);
        console.log(`Lenguaje: ${this.lenguaje}, Salario: $${this.salario}`);
        console.log(`Trabajo: ${this.trabajar()}`);
    }
}

// Uso
const original = new Desarrollador("Ana", "JavaScript", 50000);
const copia = Desarrollador.from(original);

original.mostrar();
copia.mostrar();

File: POO/Constructores_Clases/Abstract class/GO/Empleado.go
package main

import "fmt"

type Empleado interface {
    Trabajar() string
    CalcularSalario() float64
    Mostrar()
}

type Desarrollador struct {
    nombre   string
    lenguaje string
    salario  float64
}

// Constructor con parámetros
func NewDesarrollador(nombre, lenguaje string, salario float64) *Desarrollador {
    return &Desarrollador{
        nombre:   nombre,
        lenguaje: lenguaje,
        salario:  salario,
    }
}

// Constructor de copia
func NewDesarrolladorFrom(other *Desarrollador) *Desarrollador {
    return &Desarrollador{
        nombre:   other.nombre,
        lenguaje: other.lenguaje,
        salario:  other.salario,
    }
}

func (d *Desarrollador) Trabajar() string {
    return fmt.Sprintf("Programando en %s", d.lenguaje)
}

func (d *Desarrollador) CalcularSalario() float64 {
    return d.salario
}

func (d *Desarrollador) Mostrar() {
    fmt.Printf("Desarrollador: %s\n", d.nombre)
    fmt.Printf("Lenguaje: %s, Salario: $%.2f\n", d.lenguaje, d.salario)
    fmt.Printf("Trabajo: %s\n", d.Trabajar())
}

func main() {
    original := NewDesarrollador("Ana", "Go", 50000.00)
    copia := NewDesarrolladorFrom(original)
    
    var emp1 Empleado = original
    var emp2 Empleado = copia
    
    emp1.Mostrar()
    emp2.Mostrar()
}